crates/sparrow-merge/src/merge_pipeline.rs

```diff
@@ -0,0 +1,356 @@
+use std::collections::VecDeque;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+use arrow_schema::DataType;
+use error_stack::{IntoReport, IntoReportCompat, ResultExt};
+use itertools::Itertools;
+use parking_lot::Mutex;
+use sparrow_batch::Batch;
+use sparrow_physical::StepId;
+use sparrow_scheduler::{
+    InputHandles, Partition, Partitioned, Pipeline, PipelineError, Scheduler, TaskRef,
+};
+
+use crate::gather::Gatherer;
+use crate::merge::BinaryMergeInput;
+use crate::spread::Spread;
+
+/// Runs a merge operation.
+pub struct MergePipeline {
+    /// The state for each partition.
+    partitions: Partitioned<Mutex<MergePartition>>,
```

Comment:
You have two layers of mutexes which are unnecessary:

`Partitioned<Mutex<MergePartition>>` locks the merge partition, but you can also lock the `MergePartition::inputs` and `MergePartition::gatherer`. You probably want to think about how to split things up.

As a general goal: ideally the inputs would go through some kind of channel without a lock.

Reply:
So you end up with:

```
inputs: Partitioned<Channel<...>>,
state: Partitioned<Mutex<MergePartition>>,
```

and then you don't have any mutexes in the `MergePartition` -- basically, you only get the lock in `do_work` so there is no contention for the states.